web/src/js/ducks/options.js

```diff
@@ -21,12 +22,26 @@ export default function reducer(state = defaultState, action) {
                 ...action.data,
             }
 
+        case ERROR:
+            return {
+                ...state,
+                ...action.data,
+            }
+
         default:
             return state
     }
 }
 
 export function update(options) {
-    fetchApi.put('/options', options)
-    return { type: REQUEST_UPDATE }
+    let error = ''
+    fetchApi.put('/options', options).then(
+        (response) => {
+            response.text().then(errorMsg => {
+                error = errorMsg
+                console.log(error)
+            })
+        }
+    )
+    return {type: ERROR, error}
```

Comment:
I _think_ I understand what your problem is, but I'm not sure and I may have misunderstood you. Let me try. :-)

First, we have an async request here, which, once completed should dispatch an action. This doesn't work with plain redux. However, we can use [`redux-thunk`](https://github.com/gaearon/redux-thunk) (we already use at other places) to fix that. I a nutshell, redux-thunk makes 
`dispatch` available in an action creator so that we can use it asynchronously. 😃 

Let me try to sketch a very minimal example:
```javascript
export function update(options) {
    // return a function instead of an action. redux-thunk will just evaluate this right away.
    return dispatch => {
        dispatch({action: OPTION_UPDATE_START, ...});
        fetchApi.put('/options', options).then(response => {
             // maybe do nothing here because get get a state update via websockets anyways?
            dispatch({action: OPTION_UPDATE_SUCCESS, ...});
        }).catch(error => {
            dispatch({action: OPTION_UPDATE_ERROR, err: error.message})
        });
    }
}
```

It probably makes sense to have a new optionEditor reducer that contains the client-side state (the existing option reducer then only handles server-side events):

```javascript
const defaultState = {
   /* optionName -> {isUpdating, value (client-side), error} */
}

export default function reducer(state = defaultState, action) {
    switch (action.type) {

        case OPTION_UPDATE_START:
            return {
                ...state,
                [action.option]: {
                    isUpdating: true,
                    value: action.value,
                    error: false,
                }
            }

        case OPTION_UPDATE_SUCCESS:
            s = {...state};
            del s[action.option];
            return s;

        case OPTION_UPDATE_ERROR:
            return {
                ...state,
                [action.option]: {
                    ...state[action.option],
                    isUpdating: false,
                    error: action.error,
                }
            }

        default:
            return state
    }
}
```

Every row would then connect to both the server-side and the client-side state and render appropriately.
This is all sketched out very quickly and may have some fundamental flaws - does it make any sense? 😃